### PR TITLE
New Forecaster Rnn: Recurrent Neural Networks and LSTMs using Tensorflow

### PR DESCRIPTION
**Pull request by FernandoCarazoMelo**
### Description:
Introducing `ForecasterRnn`, a tensorflow-based forecaster capable of handling multi-series data & multi-step prediction. It is ready to create basic recurrent neural networks (RNN) or long short-term memory RNNs (LSTM). 
We also added `ForecasterRnn.utils.create_and_compile_model`, a flexible utility function that enables the creation of RNNs with customizable architectures. With this function, users can now easily create models with one or more recurrent and dense layers. 

### New:
- **Added ForecasterRnn:** `ForecasterRnn.ForecasterRnn`
- **Added Utility Function:** `ForecasterRnn.utils.create_and_compile_model`
  - Enables the creation of RNNs and LSTMs with flexible architectures.
  - Supports one or more dense layers, enhancing adaptability.
- **Dev notebook:** `dev/lstm_forecasting-with-lstm.ipynb` shows an example to use this module.

### CODE REVIEW

**Review by JoaquinAmatRodrigo — APPROVED** [review]

**Comment by JoaquinAmatRodrigo** [comment]
We look forward to seeing the deep learning models available in the skforecast ecosystem. Thanks @fcarazo for your contribution!

File: skforecast/ForecasterRnn/utils.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+################################################################################
+#                             skforecast.utils                                 #
+#                                                                              #
+# This work by Joaquin Amat Rodrigo and Javier Escobar Ortiz is licensed       #
```
Comment (JoaquinAmatRodrigo) [inline skforecast/ForecasterRnn/utils.py]: Replace by : This work by skforecast team is licensed under the BSD 3-Clause License.

File: skforecast/ForecasterRnn/ForecasterRnn.py [hunk 2]
```diff
@@ -0,0 +1,1102 @@
+################################################################################
+#                        ForecasterRnn                                         #
+#                                                                              #
+# This work by Joaquin Amat Rodrigo and Javier Escobar Ortiz is licensed       #
```
Comment (JoaquinAmatRodrigo) [inline skforecast/ForecasterRnn/ForecasterRnn.py]: This work by skforecast team is licensed under the BSD 3-Clause License.

**Review by JoaquinAmatRodrigo — CHANGES_REQUESTED** [review]

File: skforecast/ForecasterRnn/utils.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+# under the BSD 3-Clause License.                                              #
+################################################################################
+# coding=utf-8
+
+from typing import Union, Any, Optional, Tuple, Callable
+import pandas as pd
+
+import tensorflow as tf
+from tensorflow import keras
```
Comment (JoaquinAmatRodrigo) [inline skforecast/ForecasterRnn/utils.py:12]: We need to add tensorflow to the optional dependences

File: skforecast/ForecasterRnn/utils.py (hunk 1 above)
Comment (FernandoCarazoMelo) [reply skforecast/ForecasterRnn/utils.py]: done

File: skforecast/ForecasterRnn/ForecasterRnn.py (hunk 2 above)
Comment (FernandoCarazoMelo) [reply skforecast/ForecasterRnn/ForecasterRnn.py]: done

File: skforecast/ForecasterRnn/utils.py (hunk 3 above)
Comment (FernandoCarazoMelo) [reply skforecast/ForecasterRnn/utils.py:12]: Done. See [commit](https://github.com/JoaquinAmatRodrigo/skforecast/pull/581/commits/a3a075eb0896f0bc9606145860bfc1407d3e8afb)

Comment (FernandoCarazoMelo) [reply skforecast/ForecasterRnn/utils.py:12]: @JoaquinAmatRodrigo please, review

File: pyproject.toml [hunk 4]
```diff
@@ -47,6 +47,10 @@ sarimax = [
     "statsmodels>=0.12, <0.15",
 ]
 
+rnn = [
+    "tensorflow>=2.13.0"
```
Comment (JoaquinAmatRodrigo) [inline pyproject.toml]: Should we name this dependence deep_leerning in case we extend it to more models in the future?
We also need to fix the upper version.

**Review by JoaquinAmatRodrigo — APPROVED** [review]

File: pyproject.toml (hunk 4 above)
Comment (FernandoCarazoMelo) [reply pyproject.toml]: I initially thought that RNN was generic enough, but there are indeed other deep learning models to do forecasting that may come in the future ;)

Comment (FernandoCarazoMelo) [reply pyproject.toml]: Done in commit [requirements deep_learning](https://github.com/JoaquinAmatRodrigo/skforecast/pull/581/commits/aef17c3345a84b2f60c8f7f9c903ba32e70f8cb2)

File: skforecast/ForecasterRnn/utils.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    lags: Union[int, list],
+    steps: Union[int, list],
+    levels: Union[str, int, list] = None,
+    recurrent_layer: str = "LSTM",
+    recurrent_units: Union[int, list] = 100,
+    dense_units: list = [64],
+    activation: str = "relu",
+    optimizer: object = Adam(learning_rate=0.01),
+    loss: object = MeanSquaredError(), 
+    # compile_kwars: dict = {} # TODO. menor prioridad
```
Comment (FernandoCarazoMelo) [inline skforecast/ForecasterRnn/utils.py:30]: @JoaquinAmatRodrigo , @JavierEscobarOrtiz 

Who should take precedence, compile_kwars or the parameters inside the function?

```
optimizer: object = Adam(learning_rate=0.01),
loss: object = MeanSquaredError(),
```
If I give more priority to the parameters inside the function, the default values (e.g., `loss: object = MeanSquaredError`) will override those passed in compile_kwars, right?

Comment (JavierEscobarOrtiz) [reply skforecast/ForecasterRnn/utils.py:30]: @FernandoCarazoMelo 

Please, change `compile_kwars` to  `compile_kwargs`

Yes, give priority to the parameters inside the function. If at any time `optimizer` or `loss` are inside `compile_kwargs`, they should be overwritten with the parameter values.

Comment (JoaquinAmatRodrigo) [reply skforecast/ForecasterRnn/utils.py:30]: We can add a warning if the argument  optimizer and loss are defined  inside compile_kwargs. Indicating that it is overwritten with the argument of the function.

File: skforecast/ForecasterRnn/ForecasterRnn.py [hunk 6]
```diff
@@ -0,0 +1,1157 @@
...
+
+        for step in steps:
+            if not isinstance(step, (int, np.int64, np.int32)):
+                raise TypeError(
+                    (
+                        f"`steps` argument must be an int, a list of ints or `None`. "
+                        f"Got {type(steps)}."
+                    )
+                )
+
```
Comment (JavierEscobarOrtiz) [inline skforecast/ForecasterRnn/ForecasterRnn.py:587]: @FernandoCarazoMelo 

Review the next structure of copies in the `ForecasterAutoregMultiSeries` to adapt it for this forecaster:

https://github.com/JoaquinAmatRodrigo/skforecast/pull/592/files#diff-0a6038502cb0bf4b57cc1bf597893bf8e5ded7cc81d54411927cc9adbacaa7a6
